lib/relay-git.js

```diff
@@ -0,0 +1,115 @@
+const request = require('request');
+const { v4: uuid } = require('uuid');
+const tryJSONParse = require('./try-json-parse');
+const logger = require('./log');
+
+const GIT_CLIENT_REDIRECT_PATH = '/snykgit';
+
+module.exports = {
+  GIT_CLIENT_REDIRECT_PATH,
+  // no requestHandler here because we don't need any special
+  // HTTP-to-websocket transformation for git client requests
+  response: responseHandler,
+};
+
+// 1. Git request coming in over websocket conn (logged)
+// 2. Relay over HTTP conn to git client (logged)
+// 3. Get response over HTTP conn (logged)
+// 4. Send response over websocket conn
+function responseHandler(config, io) {
+  return (brokerToken) => (
+    { url, headers = {}, method, body = null, streamingID = '' } = {},
+    emit,
+  ) => {
+    // removing GIT_CLIENT_REDIRECT_PATH from request url;
+    if (url && url.startsWith(GIT_CLIENT_REDIRECT_PATH)) url = url.slice(GIT_CLIENT_REDIRECT_PATH.length);
+    const logContext = {
+      url,
+      method,
+      headers,
+      requestId: headers['snyk-request-id'] || uuid(),
+      streamingID,
+    };
+
+    logger.debug(logContext, 'received request over websocket connection');
+
+    const { GIT_CLIENT_URL, GIT_URL, GIT_USERNAME, GIT_PASSWORD } = config;
+    let errorStatus;
+    if (brokerToken) errorStatus = 'NO_GIT_CLIENT_ON_BROKER_SERVER';
+    else if (!GIT_CLIENT_URL) errorStatus = 'MISSING_GIT_CLIENT_URL';
+    else if (!GIT_URL || !GIT_USERNAME || !GIT_PASSWORD) errorStatus = 'MISSING_GIT_SERVER_CONFIG';
+    if (errorStatus) {
+      logger.error(
+        { ...logContext, errorStatus },
+        'received git client request with wronng configuration'
+      );
+      return emit({
+        status: 500,
+        body: errorStatus,
+      });
+    }
+
+    // injecting scm url and credentials in the request body
+    if (body) {
+      const parsedBody = tryJSONParse(body);
+      if (parsedBody && parsedBody.key && parsedBody.key.gitURI) {
+        parsedBody.key.gitURI = parsedBody.key.gitURI.includes('://') ?
```

Comment:
Question: This seems a bit shaky to me, can't we make a stronger assumption about `gitURI` here?

Reply:
Another good point. In both the git-client and DeepCode's API-server, we use `joi` to check the exact format of the `gitURI`, so the broker output value **must** be `<schema>://<scm>/<something-something>`. Now the input on the other end is controlled by registry. Right now that we only use this for Code Access, I'm ok with sending a fake `<schema>://<fake-scm>` in the request to be converted by the Broker with the correct address, but in future we might want to change it (or other products may find it obnoxious and prefer to just output `/<owner>/<repo>.git`). So the idea here is to be a bit loose on the request and tolerant with the format to leave a bit of space for further improvements without having to deploy new Broker Clients to every customer and deal with payload versioning.

Reply:
> the idea here is to be a bit loose on the request and tolerant with the format to leave a bit of space for further improvements

I see your point! 👍 Although, we can only tolerant on input as long as we can make the right assumptions without breaking things. To this point this is still fine I think.

Reply:
On a counter argument, different SCM may use different `schema` to access their git server (e.g. `ftp://`) and we could break things by injecting a default schema. Also, with no verification on the `gitURI` format, this tolerance may easily backfire. In any case, we are owning the gitURI format in the requests, so we can sync Broker and Registry accordingly.